flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/OperatorSubtaskState.java

```diff
@@ -327,4 +289,252 @@ public boolean hasState() {
+			this.inputChannelState = checkNotNull(inputChannelState);
+			return this;
+		}
+
+		public Builder setResultSubpartitionState(StateObjectCollection<ResultSubpartitionStateHandle> resultSubpartitionState) {
+			this.resultSubpartitionState = checkNotNull(resultSubpartitionState);
+			return this;
+		}
+
+		public Builder setInputChannelMappings(VirtualChannelMapping inputChannelMappings) {
+			this.inputChannelMappings = checkNotNull(inputChannelMappings);
+			return this;
+		}
+
+		public Builder setOutputChannelMappings(VirtualChannelMapping outputChannelMappings) {
+			this.outputChannelMappings = checkNotNull(outputChannelMappings);
+			return this;
+		}
+
+		public OperatorSubtaskState build() {
+			return new OperatorSubtaskState(
+				managedOperatorState,
+				rawOperatorState,
+				managedKeyedState,
+				rawKeyedState,
+				inputChannelState,
+				resultSubpartitionState,
+				inputChannelMappings,
+				outputChannelMappings);
+		}
+	}
+
+	/**
+	 * Captures ambiguous mappings of old channels to new channels.
+	 *
+	 * <p>For inputs, this mapping implies the following:
+	 * <li>
+	 *     <ul>{@link #oldTaskInstances} is set when there is a rescale on this task potentially leading to different
+	 *     key groups. Upstream task has a corresponding {@link #partitionMappings} where it sends data over
+	 *     virtual channel while specifying the channel index in the VirtualChannelSelector. This subtask then
+	 *     demultiplexes over the virtual subtask index.</ul>
+	 *     <ul>{@link #partitionMappings} is set when there is a downscale of the upstream task. Upstream task has
+	 *     a corresponding {@link #oldTaskInstances} where it sends data over virtual channel while specifying the
+	 *     subtask index in the VirtualChannelSelector. This subtask then demultiplexes over channel indexes.</ul>
+	 * </li>
+	 *
+	 * <p>For outputs, it's vice-versa. The information must be kept in sync but they are used in opposite ways for
+	 * multiplexing/demultiplexing.
+	 *
+	 * <p>Note that in the common rescaling case both information is set and need to be simultaneously used. If the
+	 * input subtask subsumes the state of 3 old subtasks and a channel corresponds to 2 old channels, then there are
+	 * 6 virtual channels to be demultiplexed.
+	 */
+	public static class VirtualChannelMapping implements Serializable {
+		public static final PartitionMapping NO_CHANNEL_MAPPING = new PartitionMapping(emptyList());
+		public static final List<PartitionMapping> NO_PARTITIONS = emptyList();
+		public static final BitSet NO_SUBTASKS = new BitSet();
+		public static final VirtualChannelMapping NO_MAPPING = new VirtualChannelMapping(NO_SUBTASKS, NO_PARTITIONS);
+
+		/**
+		 * Set when several operator instances are merged into one.
+		 */
+		private final BitSet oldTaskInstances;
+
+		/**
+		 * Set when channels are merged because the connected operator has been rescaled.
+		 */
+		private final List<PartitionMapping> partitionMappings;
```

Comment:
Isn't array and array-list the same in complexity? I usually use a map for sparse data and list when it's dense (ideally consecutive), since it's much better on the memory consumption.

Reply:
> Isn't array and array-list the same in complexity?

Yes, but here it is `List`, which can be a `LinkedList`.

If not arrays, then I'd go with Map, just because it makes obvious how elements are indexed and accessed. I don't think overhead will be visible here.